src/model_theory/basic.lean

```diff
@@ -0,0 +1,328 @@
+structure language : Type (l+1) :=
+(functions : ℕ → Type l) (relations : ℕ → Type l)
+
+namespace language
+
+/-- The empty language has no symbols. -/
+def empty : language := ⟨λ _, pempty, λ _, pempty⟩
+
+instance : inhabited language := ⟨empty⟩
+
+/-- The type of constants in a given language. -/
+@[nolint has_inhabited_instance] def const (L : language.{l}) : Type l := L.functions 0
+
+variable (L : language.{u})
+
+/-- A language is relational when it has no function symbols. -/
+def is_relational : Prop := ∀ n, L.functions n → false
+
+/-- A language is algebraic when it has no relation symbols. -/
+def is_algebraic : Prop := ∀ n, L.relations n → false
+
+variable {L}
+
+lemma is_relational_of_empty {symb : ℕ → Type u}  : is_relational ⟨λ _, pempty, symb⟩ :=
+by { intro n, apply pempty.elim }
+
+lemma is_algebraic_of_empty {symb : ℕ → Type u}  : is_algebraic ⟨symb, λ _, pempty⟩ :=
+by { intro n, apply pempty.elim }
+
+variables (L) (M : Type u)
+
+/-- A first-order structure on a type `M` consists of interpretations of all the symbols in a given
+  language. Each function of arity `n` is interpreted as a function sending tuples of length `n`
+  (modeled as `(fin n → M)`) to `M`, and a relation of arity `n` is a function from tuples of length
+  `n` to `Prop`. -/
+class Structure :=
+(fun_map : ∀{n}, L.functions n → (fin n → M) → M)
+(rel_map : ∀{n}, L.relations n → (fin n → M) → Prop)
+
+variables (N : Type u) [L.Structure M] [L.Structure N]
+
+open first_order.language.Structure
+
+/-- A homomorphism between first-order structures is a function that commutes with the
+  interpretations of functions and maps tuples in one structure where a given relation is true to
+  tuples in the second structure where that relation is still true. -/
+protected structure hom :=
+(to_fun : M → N)
+(map_fun' : ∀{n} (f : L.functions n) x, to_fun (fun_map f x) = fun_map f (to_fun ∘ x))
+(map_rel' : ∀{n} (r : L.relations n) x, rel_map r x → rel_map r (to_fun ∘ x))
+
+notation A ` →[`:25 L `] ` B := L.hom A B
+
+/-- An embedding of first-order structures is an embedding that commutes with the
+  interpretations of functions and relations. -/
+protected structure embedding extends M ↪ N :=
+(map_fun' : ∀{n} (f : L.functions n) x, to_fun (fun_map f x) = fun_map f (to_fun ∘ x))
+(map_rel' : ∀{n} (r : L.relations n) x, rel_map r (to_fun ∘ x) ↔ rel_map r x)
+
+notation A ` ↪[`:25 L `] ` B := L.embedding A B
+
+/-- An equivalence of first-order structures is an equivalence that commutes with the
+  interpretations of functions and relations. -/
+protected structure equiv extends M ≃ N :=
+(map_fun' : ∀{n} (f : L.functions n) x, to_fun (fun_map f x) = fun_map f (to_fun ∘ x))
+(map_rel' : ∀{n} (r : L.relations n) x, rel_map r (to_fun ∘ x) ↔ rel_map r x)
+
+notation A ` ≃[`:25 L `] ` B := L.equiv A B
+
+variables {L M N}
+namespace hom
+
+@[simps] instance has_coe_to_fun : has_coe_to_fun (M →[L] N) :=
+⟨(λ _, M → N), first_order.language.hom.to_fun⟩
```

Comment:
There probably should be a simp lemma that automatically rewrites `.to_hom` into a coercion. Usually we use the coercions as the simp normal form.